### ISSUE TITLE
[Pro] ~95% of response time for a large query is due to framework cpu processing

### ISSUE THREAD

**Issue by jorihardman** [rmosolgo/graphql-ruby#3379]
**Describe the bug**

I'm experiencing a performance issue which is very likely on my side, but I would appreciate some guidance in tracking it down. I have a graphql service through which users can query and locate assets that they own. I have one query in particular (shown in the graphql query section below) that is taking on average >20s to return results. We leverage graphql-batch to minimize database round trips and overall, <1s of this response time is io. The second level of the provided query, the `assets` field, returns ~3500 object and fetches nested fields for each of these objects. For a query of this size and depth, does this response time seem typical? For us, every other query to our server returns on average <1s so this definitely sticks out to us. I've attached the trace from new relic for one instance of the query. What I'm primarily interested in are the two final segments "Application code (in execute)". There is a lot of cpu burn happening in these segments and I'm curious if you have any insights into what the framework is doing at these steps. Is all of this simply ruby burn from serializing thousands of fields? I realize this background data is a bit skimpy, but I'm at a loss of where to look next to identify this cpu bottleneck. I'd be happy to provide more information.

![Screen Shot 2021-03-09 at 2 44 02 PM](https://user-images.githubusercontent.com/99236/110550070-3e42be80-80f0-11eb-806b-587d08ab8c6d.png) 

**Versions**

`graphql` version: 1.11.7
`rails` (or other framework): 5.2.4.3
`graphql-batch`: 0.4.3
`graphql-pro`: 1.16.2

**GraphQL schema**

Here is the current schema I'm using:

```ruby
class GraphqlApiSchema < GraphQL::Schema
  use GraphQL::Execution::Interpreter
  use GraphQL::Analysis::AST
  use GraphQL::Tracing::NewRelicTracing, set_transaction_name: true
  use GraphQL::Backtrace
  use GraphQL::Batch

  default_max_page_size 10000

  query Types::QueryType

  mutation Types::MutationType

  use RavenTimeout, max_seconds: ENV.fetch("REQUEST_TIMEOUT").to_i
end

```

**GraphQL query**

Example GraphQL query and response (if query execution is involved)

```graphql
query MyAssets($facilityId: ID!) {
  facility(id: $facilityId) {
    id
    assets {
      id
      assetId
      batteryStatus
      serialNumber
      name
      labels
      hwid
      manufacturer
      modelName
      ownership
      lastMovedAt
      lastSeenAt
      owners {
        id
        name
      }
      location {
        id
        name
        locationType
        unit {
          id
          name
        }
      }
      assetDeviceType: deviceType {
        id
        name
      }
      imageId
    }
  }
}
```

**Comment by jorihardman** [comment]
Seems related to https://github.com/rmosolgo/graphql-ruby/issues/2352 and https://github.com/rmosolgo/graphql-ruby/issues/861

**Comment by rmosolgo** [comment]
Hey, thanks for all the details on this, and sorry for the trouble! Are you able to replicate this slow query in development (or some environment where you can run a profiler)? If so, I have a couple of profiling suggestions. If you could share the output of the scripts below, that would inform our digging. (If you don't want to share them here, feel free to email them.)  

### Stackprof 

This will measure where Ruby is spending its time, pointing at specific methods. That will give us a clue which _part_ of the application or framework is taking so long!

```ruby
# bundle add stackprof && bundle install 
require 'stackprof'
# run the query once here 

StackProf.run(mode: :wall, out: "interpreter-profile.dump") do 
  # then run it again here
end 
```

### Memory Profiling 

This will count where object allocations are coming from. Although allocations aren't the same as cpu burn, in my experience, allocations provide a good clue to the hotspots that are actually causing a problem. 

```ruby
# bundle add memory_profiler && bundle install
require 'memory_profiler'

# first, run the query once here 

report = MemoryProfiler.report do
  # then, run the query again here 
end

report.pretty_print(to_file: "graphql_query_memory.txt")
```

------

If you're not able to replicate the issue in a profiler ... hmm, we'll figure out something. What service is that screenshot from?

**Comment by beautifulcode** [comment]
First off, thanks for your work on this project. It's great! Caveat: I'm not using Pro.
The screenshot looks like New Relic, which I also use to monitor performance
However, I am also seeing this type of perf profile, seeing large amount of time spent in "execute". 
DB is quick, I've eliminated any N+1s etc. 
I will attempt to use stackprof & memory_profiler to provide any additional info and report my findings.

**Comment by jorihardman** [comment]
Hey sorry for the slow followup and thanks for the response! It is indeed New Relic. I am able to reproduce in development, and I will try to set aside an hour this week to dump a stackprof of the request.

Unfortunately this is a critical customer-facing request for us, so for the short term, we're going to pass through this particular request to a [go implementation](https://github.com/99designs/gqlgen) of this subset of our schema. I realize that's not an ideal solution for most people here, but initial tests in the go framework (aided by batching and true io concurrency) is much much faster for us - currently <1s for the same request. That said, getting to the bottom of this execution time will likely aid us in optimizing the rest of our graphql-ruby schema, so I still very much am planning on digging into this further. Thanks for your patience!

**Comment by siav92** [comment]
Hi @rmosolgo, I'm having a similar performance problem in my application where a large number of objects in the response results in a very slow(10-15 seconds) turnaround. I used the profiling tools you suggested above and got the following:

[stackprof dump in wall, cpu and object mode](https://github.com/rmosolgo/graphql-ruby/files/6321476/stackprof.txt)
[memory profiler logs](https://github.com/rmosolgo/graphql-ruby/files/6321479/memory_profiler.txt)

I was wondering if you can help us track down the bottleneck using the logs above? Is there a way we can enhance the query performance? 

Here is the query we're making in case it helps:
[sample query](https://github.com/rmosolgo/graphql-ruby/files/6321485/sample_query.txt)


Our Stack:
````
ruby 2.7.1
rails (6.0.3.6)
graphiql-rails (1.7.0)
graphql (1.12.8)
graphql-batch (0.4.3)
graphql-pro (1.17.12)
```

Sidenote: I can make this into a separate issue if you'd prefer.

**Comment by rmosolgo** [comment]
@siav92 , I copied your comment to #3442, let's discuss there :+1:

**Comment by rmosolgo** [comment]
I just shipped 1.12.11 which included a pretty big performance improvement (#3494). If you get time, please try updating to that version! Happy to dig in further if you get any profiling results on that version.
